### ISSUE TITLE
Execution Error - Whitebox Lesson - CRSError

### ISSUE THREAD

**Issue by wbucklerbillboard** [giswqs/geog-312#151]
### Environment Information

- Python version: Python 3.12.7
- Operating System: Windows 10 Pro
- Jupyter Notebook, Anaconda Navigator and Anaconda Prompt
- Brave web browser

### Description

In the Whitebox lesson, cell 19, I executed the leafmap.get_3dep_dem function to "download a Digital Elevation Model (DEM) from 
        the USGS 3DEP Elevation service"

Following execution, CRSError

Lesson is found in https://geog-312.gishub.org/book/software/overview.html > Geopython/Whitebox (left column)

Thanks for your time!

### What I Did

Executed command at cell 19:
```
    array = leafmap.get_3dep_dem(
    gdf,
    resolution=30,
    output="dem.tif",
    dst_crs="EPSG:3857",
    to_cog=True,
    overwrite=True,
)
array
```
I also "install(ed) the required packages, such as leafmap, and whitebox. Uncomment the following lines to install the packages."

Trackback (bottommost message):
   ```
 CRSError: The EPSG code is unknown. PROJ: proj_create_from_database: C:\Users\ 
        (myname)\anaconda3\envs\geopython_QiWu\Library\share\proj\proj.db contains DATABASE.LAYOUT.VERSION.MINOR = 3 
        whereas a number >= 4 is expected. It comes from another PROJ installation.
```

Additional info used to better understand problem:
    https://github.com/geopandas/contextily/issues/174
    https://leafmap.org/workshops/ICRW_2023/?h=leafmap.get_3dep_dem (same EPSG used)

Full Trackback Message: 
[Cell19Traceback.txt](https://github.com/user-attachments/files/17723850/Cell19Traceback.txt)

**Comment by giswqs** [comment]
This is a pyproj installation issue. Try `pip install -U pyproj`

**Comment by wbucklerbillboard** [comment]
Thanks for the suggestion and swift response!

Unfortunately, it's become a stubborn issue.  I'm still getting the same `CRSError`.  Would you like me to provide an installation list, or other background info?

After running `pip install -U pyproj` and `pip install -U proj` within the Notebook, I get the following respective confirmations:

```
Requirement already satisfied: pyproj in c:\users\myname\anaconda3\envs\geopython_qiwu\lib\site-packages (3.7.0)
Requirement already satisfied: certifi in c:\users\myname\anaconda3\envs\geopython_qiwu\lib\site-packages (from pyproj) (2024.8.30)
Note: you may need to restart the kernel to use updated packages.

```

```
Requirement already satisfied: proj in c:\users\myname\anaconda3\envs\geopython_qiwu\lib\site-packages (0.2.0)
Requirement already satisfied: arrow>=0.15.6 in c:\users\myname\anaconda3\envs\geopython_qiwu\lib\site-packages (from proj) (1.3.0)
Requirement already satisfied: click>=7.1.2 in c:\users\myname\anaconda3\envs\geopython_qiwu\lib\site-packages (from proj) (8.1.7)
Requirement already satisfied: python-dateutil>=2.7.0 in c:\users\myname\anaconda3\envs\geopython_qiwu\lib\site-packages (from arrow>=0.15.6->proj) (2.9.0)
Requirement already satisfied: types-python-dateutil>=2.8.10 in c:\users\myname\anaconda3\envs\geopython_qiwu\lib\site-packages (from arrow>=0.15.6->proj) (2.9.0.20241003)
Requirement already satisfied: colorama in c:\users\myname\anaconda3\envs\geopython_qiwu\lib\site-packages (from click>=7.1.2->proj) (0.4.6)
Requirement already satisfied: six>=1.5 in c:\users\myname\anaconda3\envs\geopython_qiwu\lib\site-packages (from python-dateutil>=2.7.0->arrow>=0.15.6->proj) (1.16.0)
Note: you may need to restart the kernel to use updated packages.

```

**Comment by giswqs** [comment]
Are you using Python 3.7? If yes, try create a new env with Python 3.12.

**Comment by wbucklerbillboard** [comment]
Instead, I'm still using Python 3.12.7

**Comment by giswqs** [comment]
Create a fresh conda env to install packages. 

```python
conda create -n geo python=3.12
conda activate geo
conda install -n base mamba -c conda-forge
mamba install -c conda-forge geospatial
```

**Comment by giswqs** [comment]
See https://github.com/rasterio/rasterio/discussions/2721

**Comment by wbucklerbillboard** [comment]
Unfortunately, I've traded one issue for another.  Executing the leafmap.get_3dep_dem() function results in:
`
AttributeError: module 'leafmap' has no attribute 'get_3dep_dem'`

Before getting the new error, I created a new environment with Python 3.12 and performed the pip installs in lines 3 and 4.

After line 3 pip install:
```
WARNING: Failed to remove contents in a temporary directory 'C:\Users\myname\anaconda3\envs\geopython_QiWu312\Lib\site-packages\~umpy.libs'.
  You can safely remove it manually.
  WARNING: Failed to remove contents in a temporary directory 'C:\Users\myname\anaconda3\envs\geopython_QiWu312\Lib\site-packages\~umpy'.
  You can safely remove it manually.
```

After line 4 pip install:
```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pygeoutils 0.18.0 requires numpy>=2, but you have numpy 1.26.4 which is incompatible.
pynhd 0.18.0 requires numpy>=2, but you have numpy 1.26.4 which is incompatible.
```

I looked at the suggested rasterio link, which was informative.

Is there something else I can try, please?

**Comment by giswqs** [comment]
Check your leafmap version. Make sure you install the latest version v0.40.1

**Comment by wbucklerbillboard** [comment]
I updated leafmap to v0.40.1 (via `pip install -U XXXXX`)

However, I ran into another issue.  Upon executing gdf.explore() (Cell 15), I ran into the following error message:

`ImportError: The 'folium', 'matplotlib' and 'mapclassify' packages are required for 'explore()'. You can install them using 'conda install -c conda-forge folium matplotlib mapclassify' or 'pip install folium matplotlib mapclassify'.`

I also executed the two pips installs in the first two cells.

I've enclosed my python package listing and full extent of the ImportError:
[PackList&ExploreFunc_ImportError.txt](https://github.com/user-attachments/files/17805369/PackList.ExploreFunc_ImportError.txt)

Thanks again!

**Comment by giswqs** [comment]
`pip install mapclassify`

**Comment by wbucklerbillboard** [comment]
I still have the same ImportError.  After performing `pip install mapclassify` in Anaconda Prompt and Notebook.

Additionally, as I run `leafmap.get_3dep_dem`, a message begins with `py3dep is not installed. Installing py3dep` and ends with `ImportError: DLL load failed while importing _pocketfft_umath: The specified module could not be found.`

**Comment by wbucklerbillboard** [comment]
Should I try something different, please?

**Comment by giswqs** [comment]
Try using Google Colab if Windows installation does not work for you. This is an issue with packages specific to Windows, not really an issue with this course. I don't use Windows, and I can't test all package installations on Windows. Some Python packages are notoriously difficult to install on Windows. You will have to use Google to find answers unless it is a common issue that I also run into on Linux.

**Comment by wbucklerbillboard** [comment]
You're right, Colab is working better for me.  Thanks!

I learned a great deal!  Even noting how mundane aspects like OS, packages, versions and web browser environments can combine to confound.

While I am farther along in the Whitebox lesson, I've hit a snag within its Colab form.  At Add Color Bar / `leafmap.image_min_max("dem.tif")` I'm getting:

`AttributeError: module 'leafmap' has no attribute 'image_min_max'`

I used  `%pip install leafmap --upgrade` and my Leafmap's version is 0.41.2 (before the above error)

Should I downgrade?

Thanks.
